heartbeat/monitors/active/http/check.go

```diff
@@ -132,14 +134,100 @@ func checkHeaders(headers map[string]string) respValidator {
 	}
 }
 
-func checkBody(matcher []match.Matcher) bodyValidator {
-	return func(r *http.Response, body string) error {
-		for _, m := range matcher {
-			if m.MatchString(body) {
+func parseBody(b interface{}) ([]match.Matcher, []match.Matcher) {
+	var (
+		pm, nm []match.Matcher
+		positive = "positive"
+		negative = "negative"
+	)
+	// run through this code block if there is no positive or negative keyword in response body
+	// in this case, there's only plain body
+	p, ok := b.([]interface{})
+	if ok {
+		for _, pp := range p {
+			pat, ok := pp.(string)
+			if ok {
+				return append(pm, match.MustCompile(pat)), nm
+			}
+		}
+	}
+
+	// run through this part if there exists positive/negative keyword in response body
+	// in this case, there will be 3 possibilities: positive + negative / positive / negative
+	m, ok := b.(map[string]interface{})
+	if ok {
+		for checkType, v := range m {
+			params, ok := v.([]interface{})
+			if ok {
+				for _, param := range params {
+					pat, ok := param.(string)
+					if ok {
+						if checkType == positive {
+							pm = append(pm, match.MustCompile(pat))
+						} else if checkType == negative {
+							nm = append(nm, match.MustCompile(pat))
+						} else {
+							return pm, nm
+						}
+					}
+				}
+			}
+		}
+	}
+	return pm, nm
+}
+
+/* checkBody accepts 2 parameters:
+1. positive check regex
+2. negative check regex
+*/
+func checkBody(pm, nm []match.Matcher) bodyValidator {
+	if len(pm) <= 0 && len(nm) <= 0 {
```

Comment:
I see the intent here, using the `if/else` branches to return an optimal function, but I'm not sure the tradeoff in readability / fragility of code is worth it.

I'd prefer to either only run the final case with the two `for` loops OR to build up an array of function pointers, with each one corresponding to a positive or negative match and iterate over that.

That said, we're really getting into micro optimizations here, and I'm not convinced that will yield much in this case, and I don't think it's worth the effort to benchmark, so I think the two for loops makes the most sense.

Reply:
Yea, I made things complex.
Refactored this function as you suggested, please help to review.